packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp

```diff
@@ -479,6 +478,103 @@ void Add(
+    KOKKOS_INLINE_FUNCTION void operator()(const size_t i, size_t& update, const bool finalPass) const
+    {
+      // exclusive scan
+      if(finalPass)
+      {
+        gtol(i) = update;
+      }
+      if(entryUnion(i))
+      {
+        update++;
+      }
+      if(finalPass)
+      {
+        if(i == entryUnion.dimension_0() - 1)
+        {
+          gtol(i + 1) = update;
+        }
+      }
+    }
+    const ByteView& entryUnion;
+    LView& gtol;
+  };
+
+  template<typename LView, typename GView>
+  struct ConvertGlobalToLocal
+  {
+    ConvertGlobalToLocal(const LView& gtol_, const GView& gids_, LView& lids_) : gtol(gtol_), gids(gids_), lids(lids_) {}
+    KOKKOS_INLINE_FUNCTION void operator()(const size_t i) const
+    {
+      lids(i) = gtol(gids(i));
+    }
+    const LView& gtol;
+    const GView& gids;
+    LView& lids;
+  };
+}
+
+//Build the minimal (sorted) column map for the given set of global columns
+//Then convert gids and store them in lids (gids is not modified)
+template<typename Scalar, typename LocalOrdinal, typename GlobalOrdinal, typename Node>
+Teuchos::RCP<Tpetra::Map<LocalOrdinal, GlobalOrdinal, Node>> AddKernels<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+getOptimizedColMap(GlobalOrdinal ncols,
+                   const typename AddKernels<Scalar, LocalOrdinal, GlobalOrdinal, Node>::GlobalColIndsArray& gids,
+                   typename AddKernels<Scalar, LocalOrdinal, GlobalOrdinal, Node>::ColIndsArray& lids,
+                   const Teuchos::RCP<const Teuchos::Comm<int>>& comm)
+{
+  using namespace ColMapFunctors;
+  using Teuchos::RCP;
+  using Teuchos::rcp;
+  typedef Kokkos::View<char*, device_type> ByteView;
+  typedef typename AddKernels<Scalar, LocalOrdinal, GlobalOrdinal, Node>::GlobalColIndsArray GView;
+  typedef typename AddKernels<Scalar, LocalOrdinal, GlobalOrdinal, Node>::ColIndsArray LView;
+  //Functors (explained in the procedural code below)
+  auto nentries = gids.dimension_0();
+  //each entry of entryUnion is 0 unless there is a local entry in that column (then it is 1)
+  ByteView entryUnion("entry union", ncols);
+  UnionEntries<ByteView, GView> ue(entryUnion, gids);
+  Kokkos::parallel_for(range_type(0, nentries), ue);
+  //turn entryUnion into prefix sum gtol (where gtol(i) gives the new local col for global col i)
+  LView gtol("global col -> local col", ncols + 1);
+  EntriesPrefixSum<ByteView, LView> ps(entryUnion, gtol);
+  Kokkos::parallel_scan(range_type(0, ncols), ps);
+  //convert gids to local ids and put them in lids (implicitly sorted as long as input gids is sorted per row)
+  ConvertGlobalToLocal<LView, GView> cgtl(gtol, gids, lids);
+  Kokkos::parallel_for(range_type(0, gids.dimension_0()), cgtl);
+  //build local set of GIDs for constructing column map - the last entry in gtol is the total number of local cols
+  GView colmap("column map", gtol(gtol.dimension_0() - 1));
+  size_t localIter = 0;
+  for(size_t i = 0; i < entryUnion.dimension_0(); i++)
```

Comment:
Also, could we reformulate this sequential host kernel as a `parallel_scan`?  We could have counted the total number of nonzero entries in `entryUnion`, when building `entryUnion` above.  I need to think about how this scan would look.